src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -108,67 +108,90 @@ class TestTargetSetup:
-            additional_args=additional_args_for_pytest,
-        )
+    create_requirements_pex_request = CreatePexFromTargetClosure(
+        addresses=test_addresses,
+        output_filename="requirements.pex",
+        include_source_files=False,
+        additional_args=additional_args_for_pytest,
     )
 
-    test_runner_pex = await Get[Pex](
-        CreatePex,
-        create_pex(
-            output_filename="test_runner.pex",
-            entry_point="pytest:main",
-            interpreter_constraints=interpreter_constraints,
-            additional_args=(
-                "--pex-path",
-                ":".join(
-                    pex_request.output_filename
-                    # TODO(John Sirois): Support shading python binaries:
-                    #   https://github.com/pantsbuild/pants/issues/9206
-                    # Right now any pytest transitive requirements will shadow corresponding user
-                    # requirements which will lead to problems when APIs that are used by either
-                    # `pytest:main` or the tests themselves break between the two versions.
-                    for pex_request in (pytest_pex, requirements_pex)
-                ),
+    create_test_runner_pex = create_pex(
+        output_filename="test_runner.pex",
+        entry_point="pytest:main",
+        interpreter_constraints=interpreter_constraints,
+        additional_args=(
+            "--pex-path",
+            # TODO(John Sirois): Support shading python binaries:
+            #   https://github.com/pantsbuild/pants/issues/9206
+            # Right now any pytest transitive requirements will shadow corresponding user
+            # requirements which will lead to problems when APIs that are used by either
+            # `pytest:main` or the tests themselves break between the two versions.
+            ":".join(
+                (
+                    create_pytest_pex_request.output_filename,
+                    create_requirements_pex_request.output_filename,
+                )
             ),
         ),
     )
 
-    chrooted_sources = await Get[ChrootedPythonSources](HydratedTargets(all_targets))
+    # Get the file names for the test_target so that we can specify to Pytest precisely which files
+    # to test, rather than using auto-discovery.
+    specified_source_files_request = SpecifiedSourceFilesRequest(
+        [adaptor_with_origin], strip_source_roots=True
+    )
+
+    requests: List[Get[Any]] = [
+        Get[Pex](CreatePex, create_pytest_pex_request),
+        Get[Pex](CreatePexFromTargetClosure, create_requirements_pex_request),
+        Get[Pex](CreatePex, create_test_runner_pex),
+        Get[ChrootedPythonSources](HydratedTargets(all_targets)),
+        Get[SourceFiles](SpecifiedSourceFilesRequest, specified_source_files_request),
+    ]
+    if run_coverage:
+        requests.append(
+            Get[Coveragerc](CoveragercRequest(HydratedTargets(all_targets), test_time=True)),
+        )
+
+    (
```

Comment:
I can understand why this is worthwhile and reasonable, but I nonetheless feel that adding a comment here to apologize-for/explain it would be worthwhile =)

Reply:
I won't apologize since the APIs force this upon me and others right now, but I did not we can do better and  TODO'd cleanup in #9294.

Reply:
Thanks a ton!